### PR TITLE
Add stubs package for lxml.

### PR DESCRIPTION
**Pull request by clokep**
Turns out lxml has official type hint stubs: https://github.com/lxml/lxml-stubs

See inline comments.

### CODE REVIEW

File: synapse/media/oembed.py [hunk 1]
```diff
@@ -109,18 +109,20 @@ def autodiscover_from_html(self, tree: "etree.Element") -> Optional[str]:
             The URL to use for oEmbed information, or None if no URL was found.
         """
         # Search for link elements with the proper rel and type attributes.
-        for tag in tree.xpath(
-            "//link[@rel='alternate'][@type='application/json+oembed']"
+        for tag in cast(
+            List["etree._Element"],
+            tree.xpath("//link[@rel='alternate'][@type='application/json+oembed']"),
         ):
```
Comment (clokep) [inline synapse/media/oembed.py:115]: See https://lxml.de/xpathxslt.html#xpath-return-values

tl;dr `xpath` returns a massive `Union` of types, what type is returned is determined by the expression itself. We could use asserts instead of casts though if that seems better.

File: synapse/media/oembed.py [hunk 2]
```diff
@@ -109,18 +109,20 @@ def autodiscover_from_html(self, tree: "etree.Element") -> Optional[str]:
...
             if "href" in tag.attrib:
-                return tag.attrib["href"]
+                return cast(str, tag.attrib["href"])
```
Comment (clokep) [inline synapse/media/oembed.py:117]: `attrib` is a [special `_Attrib` class](https://github.com/lxml/lxml-stubs/blob/f0290b545e2c694c66934ec36b6e77b990cceb1e/lxml-stubs/etree.pyi#L301-L331), the `__getitem__` takes an `_AnyStr` and returns an `_AnyStr`, I think this works if you give a string you get a string (and if you give bytes, you get bytes). Anyway, we always get a `str` back but mypy thinks it could also be `bytes`.

File: synapse/media/oembed.py [hunk 3]
```diff
@@ -244,11 +246,12 @@ def calc_description_and_urls(open_graph_response: JsonDict, html_body: str) ->
...
 
     # Attempt to parse the body. If this fails, log and return no metadata.
-    tree = etree.fromstring(html_body, parser)
+    # TODO Develop of lxml-stubs has this correct.
+    tree = etree.fromstring(html_body, parser)  # type: ignore[arg-type]
 
     # The data was successfully parsed, but no tree was found.
     if tree is None:
-        return
+        return  # type: ignore[unreachable]
```
Comment (clokep) [inline synapse/media/oembed.py:254]: `fromstring` is defined as always returning `_Element`, but empirically it also returns `None`.

File: synapse/media/preview_html.py [hunk 4]
```diff
@@ -381,8 +405,8 @@ def parse_html_description(tree: "etree.Element") -> Optional[str]:
 
 
 def _iterate_over_text(
-    tree: Optional["etree.Element"],
-    tags_to_ignore: Set[Union[str, "etree.Comment"]],
+    tree: Optional["etree._Element"],
+    tags_to_ignore: Set[object],
```
Comment (clokep) [inline synapse/media/preview_html.py:409]: `etree.Comment` isn't a type (since it is a function), and (as the comment above states) using `etree._Comment` is not the correct thing to do.

**Comment by clokep** [comment]
> I think I previously looked at using this stubs package before---at least, enough for me to raise [lxml/lxml-stubs#63](https://github.com/lxml/lxml-stubs/issues/63) and [lxml/lxml-stubs#64](https://github.com/lxml/lxml-stubs/pull/64). I would guess that the stubs are more complete these days?

They haven't done a release in a while so (since March of 2022) so I don't think your improvements are in there.

I'm not sure if the amount of casts and such makes this worthwhile, but I like having it typed.

File: synapse/media/oembed.py [hunk 5]
```diff
@@ -98,7 +98,7 @@ def get_oembed_url(self, url: str) -> Optional[str]:
         # No match.
         return None
 
-    def autodiscover_from_html(self, tree: "etree.Element") -> Optional[str]:
+    def autodiscover_from_html(self, tree: "etree._Element") -> Optional[str]:
```
Comment (DMRobertson) [inline synapse/media/oembed.py:101]: What's the difference between `Element` and `_Element`? Perhaps the annotation should have always been `_Element`, and `_Element` isn't publicly exported?

File: synapse/media/preview_html.py [hunk 6]
```diff
@@ -247,7 +255,7 @@ def parse_html_to_open_graph(tree: "etree.Element") -> Dict[str, Optional[str]]:
     # "og:video:height" : "720",
     # "og:video:secure_url": "https://www.youtube.com/v/LXDBoHyjmtw?version=3",
 
-    og = _get_meta_tags(tree, "property", "og")
+    og: Dict[str, Optional[str]] = _get_meta_tags(tree, "property", "og")
```
Comment (DMRobertson) [inline synapse/media/preview_html.py:258]: I guess we need to explicitly annotate this because it is a recursive self-call?

**Review by DMRobertson — APPROVED** [review]
Thanks for your thoroughness here.

File: synapse/media/oembed.py (hunk 5 above)
Comment (clokep) [reply synapse/media/oembed.py:101]: `Element` is a function which creates an `_Element`:

```pycon
>>> from lxml import etree
>>> etree.Element
<cyfunction Element at 0x10d478c70>
>>> etree._Element
<class 'lxml.etree._Element'>
```

File: synapse/media/preview_html.py (hunk 6 above)
Comment (clokep) [reply synapse/media/preview_html.py:258]: I actually don't think we need this anymore based on backing out other changes.

File: synapse/media/oembed.py (hunk 1 above)
Comment (clokep) [reply synapse/media/oembed.py:115]: For every single instance?

File: synapse/media/oembed.py (hunk 5 above)
Comment (DMRobertson) [reply synapse/media/oembed.py:101]: Aha!

File: synapse/media/oembed.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/media/oembed.py:116]: Oh, sorry---there are quite a few of them. In that case it's probably fine to leave it uncommented (I guess our future selves can always rediscover this PR.) 

(Though it seems I responded too slowly. I'll let you keep or remove the comments at your discretion.)
